infrastructure/w3id/src/logs/log-manager.ts

```
@@ -0,0 +1,148 @@
+import type { StorageSpec } from "./storage/storage-spec";
+
+/**
+ * Class to generate historic event logs for all historic events for an Identifier
+ * starting with generating it's first log entry
+ */
+
+// TODO: Create a specification link inside our docs for how generation of identifier works
+
+export class IDLogManager {
+	repository: StorageSpec<LogEvent, LogEvent>;
+
+	constructor(repository: StorageSpec<LogEvent, LogEvent>) {
+		this.repository = repository;
+	}
+
+	static async validateLogChain(
+		log: LogEvent[],
+		verifyCallback: VerifierCallback,
+	) {
+		let currIndex = 0;
+		let currentNextKeyHashesSeen: string[] = [];
+		let lastUpdateKeysSeen: string[] = [];
+		let lastHash: string | null = null;
+
+		for (const e of log) {
+			const [_index, _hash] = e.versionId.split("-");
+			const index = Number(_index);
+			if (currIndex !== index) throw new MalformedIndexChainError();
+			const hashedUpdateKeys = await Promise.all(
+				e.updateKeys.map(async (k) => await hash(k)),
+			);
+			if (index > 0) {
+				const updateKeysSeen = isSubsetOf(
+					hashedUpdateKeys,
+					currentNextKeyHashesSeen,
+				);
+				if (!updateKeysSeen || lastHash !== _hash)
+					throw new MalformedHashChainError();
+			}
+
+			currentNextKeyHashesSeen = e.nextKeyHashes;
+			await IDLogManager.verifyLogEventProof(
+				e,
+				lastUpdateKeysSeen.length > 0 ? lastUpdateKeysSeen : e.updateKeys,
+				verifyCallback,
+			);
+			lastUpdateKeysSeen = e.updateKeys;
+			currIndex++;
+			lastHash = await hash(canonicalize(e) as string);
+		}
+		return true;
+	}
+
+	private static async verifyLogEventProof(
+		e: LogEvent,
+		currentUpdateKeys: string[],
+		verifyCallback: VerifierCallback,
+	) {
+		const proof = e.proof;
+		const copy = JSON.parse(JSON.stringify(e));
+		// biome-ignore lint/performance/noDelete: we need to delete proof completely
+		delete copy.proof;
+		const canonicalJson = canonicalize(copy);
+		let verified = false;
+		if (!proof) throw new Error();
```

lets throw a specific error too?

---

fixed https://github.com/MetaState-Prototype-Project/prototype/pull/98/commits/d1de58a305e44bd6f3aadbb6e4639d20f82f4497